src/wayland/shell/xdg/mod.rs

```diff
@@ -175,88 +427,150 @@ impl PositionerState {
+                    || self.anchor_edges == xdg_positioner::Anchor::TopRight
+            }
+            xdg_positioner::Anchor::Bottom => {
+                self.anchor_edges == xdg_positioner::Anchor::Bottom
+                    || self.anchor_edges == xdg_positioner::Anchor::BottomLeft
+                    || self.anchor_edges == xdg_positioner::Anchor::BottomRight
+            }
+            xdg_positioner::Anchor::Left => {
+                self.anchor_edges == xdg_positioner::Anchor::Left
+                    || self.anchor_edges == xdg_positioner::Anchor::TopLeft
+                    || self.anchor_edges == xdg_positioner::Anchor::BottomLeft
+            }
+            xdg_positioner::Anchor::Right => {
+                self.anchor_edges == xdg_positioner::Anchor::Right
+                    || self.anchor_edges == xdg_positioner::Anchor::TopRight
+                    || self.anchor_edges == xdg_positioner::Anchor::BottomRight
+            }
+            _ => unreachable!(),
+        }
+    }
+
+    pub(crate) fn gravity_has_edge(&self, edge: xdg_positioner::Gravity) -> bool {
+        match edge {
+            xdg_positioner::Gravity::Top => {
+                self.gravity == xdg_positioner::Gravity::Top
+                    || self.gravity == xdg_positioner::Gravity::TopLeft
+                    || self.gravity == xdg_positioner::Gravity::TopRight
+            }
+            xdg_positioner::Gravity::Bottom => {
+                self.gravity == xdg_positioner::Gravity::Bottom
+                    || self.gravity == xdg_positioner::Gravity::BottomLeft
+                    || self.gravity == xdg_positioner::Gravity::BottomRight
+            }
+            xdg_positioner::Gravity::Left => {
+                self.gravity == xdg_positioner::Gravity::Left
+                    || self.gravity == xdg_positioner::Gravity::TopLeft
+                    || self.gravity == xdg_positioner::Gravity::BottomLeft
+            }
+            xdg_positioner::Gravity::Right => {
+                self.gravity == xdg_positioner::Gravity::Right
+                    || self.gravity == xdg_positioner::Gravity::TopRight
+                    || self.gravity == xdg_positioner::Gravity::BottomRight
+            }
+            _ => unreachable!(),
+        }
+    }
 }
 
-/// Contents of the pending state of a shell surface, depending on its role
-#[derive(Debug)]
-pub enum XdgSurfacePendingState {
-    /// This a regular, toplevel surface
-    ///
-    /// This corresponds to the `xdg_toplevel` role
-    ///
-    /// This is what you'll generally interpret as "a window".
-    Toplevel(ToplevelState),
-    /// This is a popup surface
-    ///
-    /// This corresponds to the `xdg_popup` role
-    ///
-    /// This are mostly for small tooltips and similar short-lived
-    /// surfaces.
-    Popup(PopupState),
-    /// This surface was not yet assigned a kind
-    None,
+impl From<&PositionerState> for Rectangle {
```

Comment:
Could you add a comment explaining what this conversion does exactly?

Given it's a pretty non-obvious conversion, it should probably be a method of `PositionerState`, rather than a `From<_>` impl.

Reply:
I added the relevant parts of the protocol specification to the function.